Sources/SwiftFusion/Inference/Errors.swift

```diff
@@ -18,43 +18,36 @@ import TensorFlow
 public typealias Error = Vector
 
 /// Collection of all errors returned by a Factor Graph
-public typealias Errors = Array<Error>
+public struct Errors {
+  public var values: Array<Error>.DifferentiableView
+
+  /// Creates empty `Errors`.
+  public init() {
+    self.values = Array.DifferentiableView()
+  }
+
+  /// Creates `Errors` containing the given `errors`.
+  public init(_ errors: [Error]) {
+    self.values = Array.DifferentiableView(errors)
+  }
+
+  public static func += (_ lhs: inout Self, _ rhs: [Error]) {
+    lhs.values.base += rhs
+  }
+}
 
 /// Extending Array for Error type
 /// This simplifies the implementation for `Errors`, albeit in a less self-contained manner
 /// TODO: change this to a concrete `struct Errors` and implement all the protocols
-extension Array where Element == Error {
-  public static func - (_ a: Self, _ b: Self) -> Self {
-    var result = a
-    let _ = result.indices.map { result[$0] = a[$0] + b[$0] }
-    return result
-  }
-  
+extension Errors: EuclideanVectorSpace {
+
+  // Note: Requirements of `Differentiable`, `AdditiveArithmetic`, and `VectorProtocol` are automatically
+  // synthesized. Yay!
+
   /// Calculates the L2 norm
-  public var norm: Double {
+  public var squaredNorm: Double {
     get {
-      self.map { $0.squared().sum() }.reduce(0.0, { $0 + $1 })
+      values.base.map { $0.squared().sum() }.reduce(0.0, { $0 + $1 })
     }
   }
-  
-  /// Errors + scalar
-  static func + (_ lhs: Self, _ rhs: Double) -> Self {
```

Comment:
Could you explain briefly how these are automatically implemented? Curious on the workings...

Reply:
Yeah! We have automatic synthesis for `AdditiveArithmetic` and `VectorProtocol` that looks at all the fields of the struct. If all the fields conform to `AdditiveArithmetic` or `VectorProtocol`, it automatically implements the requirements by applying the functions to all the members.

Reply:
The single field `Array<Error>.DifferentiableView` conforms to `AdditveArithmetic` and `VectorProtocol` because:
1. `Error` (aka `Vector`) conforms to them.
2. `Array.DifferentiableView` conforms to `AdditiveArithmetic` when its `Element` does: https://github.com/apple/swift/blob/4bc72aedb2b32c33ed8e2ec241615fc890b60002/stdlib/public/Differentiation/ArrayDifferentiation.swift#L98
3. `Array.DifferentiableView` conforms to `VectorProtocol` when its `Element` does: https://github.com/tensorflow/swift-apis/blob/a8a24c46e478ce50c1c9a7718a41eec453e5b670/Sources/TensorFlow/StdlibExtensions.swift#L224